File: app/workshop/01_NodePy.js
#!/usr/bin/env node

// STILL NEEDS: examples, full explanation

/* Python Shell: https://www.npmjs.com/package/python-shell */
/*
 * Documentation (https://www.npmjs.com/package/python-shell#api-reference)
 * - Options
 *   - args: array of arguments, effectively sys.argv---DOES SCRIPT NAME COUNT??
 *   - scriptPath: string, path prefix for accessing scripts, if necessary
 *   - pythonPath: string, python executable, if not specified in the script
 *   - pythonOptions: array of command-line options for the executable
 *   - mode: string text/json/binary: data mode, in text, json, or terminal std
 *     - encoding: string, text encoding
 *     - formatter: method for formatting transmitted data
 *     - parser: method acting on and emitting transmitted data
 *     - stderrParser: method acting on and emitting transmitted logs (stderr)
 *
 * - PythonShell.run[Script](script/code, options, callback)
 *   - Synchronous function that runs Python and returns errors or results
 *   - script/code: a string, indicating the code that is to be run. run
 *     accesses the file by that nume; runString runs the code
 *   - options: see "options" above
 *   - callback: the callback function, running when the script is complete
 *
 * - new PythonShell(script, options)
 *   - Object Constructor: Useful but complicated
 *
 */

console.log('Welcome to the PythonShell demonstration script!\n');

/* Setup */
// IMPORT: PythonShell is a class within the `python-shell` package
const PythonShell = require('python-shell').PythonShell;

// Displaying the `console.log` for more details
console.log('Displaying the `PythonShell` class, for more information')
console.log('-', PythonShell);
console.log();

/* Basic */
// DEMO 1: Running runString as a demonstrative method
PythonShell.runString("x = 2 + 2; print(x); print(2*x)", null, (err, res) => {
  if (err) { throw err; }
  console.log('Results of `runString` method');
  console.log('-', res);
  console.log();
});

/* Standard */
// Ideally, insert an example showing specifity of programming language
// DEMO 3: "BASIC TEXT", the most basic run demonstration
PythonShell.run('01_NodePy.py', null, (err, res) => {
  if (err) { throw err; }
  console.log('[BASIC TEXT] Results of `run` method');
  console.log('- Returns an array of strings, separated at newlines');
  console.log('-', res);
  console.log('- For complex data, strings must be manually processed');
  console.log();
});

// DEMO 4: "ARGUMENTS I", demonstrating some basic flags
PythonShell.run('01_NodePy.py', {args: ['-b', '-s']}, (err, res) => {
  if (err) { throw err; }
  console.log('[ARGUMENTS I] Running with simple flags: 01_NodePy.py -b -s');
  console.log('- `-b` deactivates basic text; `-s` activates sample flag');
  console.log('- Order doesn\'t matter, as these are just flags');
  console.log('-', res);
  console.log();
});

// DEMO 5: "ARGUMENTS II", demonstrating flags with arguments
let options = {
  args: [
    'foo bar',
    '-b', '-a',
    '-one', 'three',
    '-two', 'forty', 'two',
    '-n', 'foo', 'bar',
    '--manipulate', 'foo.bar'
  ]
}
PythonShell.run('01_NodePy.py', options, (err, res) => {
  if (err) { throw err; }
  console.log('[ARGUMENTS II]: Running with a laundry list of arguments');
  console.log(`- 01_NodePy.py ${options.args}`);
  console.log('- Order doesn\'t matter greatly; see argparse for details');
  console.log('- Note that output is still just text; input is only strings:');
  console.log('-', res);
  console.log('- Observe that arguments can have spaces and be one unit');
  console.log('  and that multiple entries for an argument must be separate');
  console.log();
});

// DEMO 6: "JSON", because maybe what we have to do is print entirely in JSON?
PythonShell.run('01_NodePy.py', {args: ['--json'], mode: 'json'}, (err, res) => {
  if (err) { throw err; }
  console.log('[JSON] Running with simple flags: 01_NodePy.py --json');
  console.log("- The `mode: 'json'` option causes PythonShell to try to");
  console.log("  parse all printed output as JSON - so Python must output");
  console.log("  only JSON (see the Python Example + python.json for details)");
  console.log('-', res);
  console.log();
});

// Perhaps something with more options - show the default options?
// IF THERE'S TIME: Experiment more with stdin/stdout
   // That might be better-suited for Advanced, though

/* Advanced */
// let shell = new PythonShell('01_NodePy.py', {mode: "text"});
// console.log(shell.stderr.text)
// console.log('As you can see, the script fails quite dramatically');

File: app/workshop/01_NodePy.py
#!/usr/bin/env python3

# STILL NEEDS: JSON output, stdin/stdout, stderr
# - Make sure that this is an *comprehensive* example of cross-app communication

import sys
import argparse
import json

""""""
"""Setting up argparse: https://docs.python.org/3/library/argparse.html"""
# Initializing the Parser Object
parser = argparse.ArgumentParser(
    description='A comprehensive Python3 I/O example')

# Demo: Setting up a simple flag - in this case, (de)activating output
parser.add_argument('-b', '--no-basic', action='store_const', const=True,
                    default=False, help='deactivate basic text output')
parser.add_argument('-s', '--sample', action='store_const', const=True,
                    default=False, help='activate to test a sample flag')
parser.add_argument('-a', '--arguments', action='store_true',
                    help='activate the arguments ii demo')

# Demo: Setting up an argument - many options; note, variable is always a list
parser.add_argument('positional', action='store', nargs='*', default=[],
                    help="a sample positional argument; a sample `nargs='*'``")
parser.add_argument('-one', '--one-arg', action='store', nargs=1,
                    help="a demonstration of a flag with a single argument")
parser.add_argument('-two', '--two-arg', action='store', nargs=2,
                    help="a demonstration of a flag with two arguments")
parser.add_argument('-n', '--n-arg', action='store', nargs='+',
                    help="a demonstration of a flag `nargs='+'`")
parser.add_argument('--manipulate', action='store', nargs=1, default='',
                    type=str, help='sample text manipulation: splitting by `.`')

# Adding a JSON flag to active JSON Output
parser.add_argument('--json', action='store_true',
                    help='activate the JSON demo')

# And Lastly, Parsing the Args - Except the Script Name, Obviously
argv = parser.parse_args(sys.argv[1:])

if argv.json:
    argv.no_basic = True
    argv.sample = False
    argv.arguments = False

""""""
# BASIC TEXT: With no arguments, output a few demonstrative lines
if not(argv.no_basic):
    print(42 + 27)

    some_list = ['foo', 'bar', 'hello world']

    for item in some_list: print(item)
    print(some_list)

# ARGUMENTS I: If the sample flag is received, say so
if argv.sample:
    print('Sample flag received and processed!')
    print('Original Input: {}'.format(sys.argv))
    print('Processed Args: {}'.format(argv))

# ARGUMENTS II: Working with arguments that store input values
if argv.arguments:
    print('Argument Demo Request Received!')
    print('Positional Argument(s): {}'.format(argv.positional))
    print('Single Argument: {}; Double Argument: {}'.format(argv.one_arg,
        argv.two_arg))
    print('n-arg Argument: {}'.format(argv.n_arg))
    print('Manipulation: {}'.format(argv.manipulate[0].split('.')))

# JSON: Sample JSON Output
if argv.json:
    sample_json_dic = {
        'int': 420,
        'float': 1.618,
        'string': "hello world",
        'list': [420, 1.618, "hello world"],
        # 'function': lambda x: x + 5, - fails
    }
    print(json.dumps(sample_json_dic))
